src/RewardsStreamer.sol

```diff
@@ -77,18 +118,35 @@ contract RewardsStreamer is ReentrancyGuard {
         user.userRewardIndex = rewardIndex;
     }
 
-    function updateRewardIndex() public {
-        if (totalStaked == 0) {
+    /**
+     * @dev Updates the reward index and accounted rewards based on the current
+     * block number. This function is called before any state-modifying operations
+     * to ensure that the reward calculations are up-to-date. It updates the
+     * `rewardIndex` to reflect the latest calculated value and increments
+     * `accountedRewards` with the new rewards accrued since the last update.
+     */
+    function updateRewardIndex() internal {
+        if (totalStaked == 0 || block.number >= rewardEndBlock) {
+            lastRewardBlock = block.number;
             return;
         }
 
-        uint256 rewardBalance = REWARD_TOKEN.balanceOf(address(this));
-        uint256 newRewards = rewardBalance > accountedRewards ? rewardBalance - accountedRewards : 0;
+        uint256 blocksSinceLastUpdate = block.number - lastRewardBlock;
+        uint256 applicableBlocks = blocksSinceLastUpdate;
 
-        if (newRewards > 0) {
-            rewardIndex += (newRewards * SCALE_FACTOR) / totalStaked;
-            accountedRewards += newRewards;
+        if (block.number > rewardEndBlock) {
+            applicableBlocks = rewardEndBlock - lastRewardBlock;
```

Comment:
I think this is problematic because `rewardEndBlock` is not set during construction. Meaning, even when `setRewardsPerBlock` is called the first time, `rewardEndBlock` will be `0` if I'm not mistaken.

Reply:
Good point, gotta admit I kind of neglected the initiation and edge cases 😅